src/amici.cpp

```diff
@@ -1383,6 +1384,375 @@ int getDiagnosisB(int it, void *ami_mem, UserData *udata, ReturnData *rdata, Tem
+        res_old = sqrt(res_old);
+        
+        // if the step was useful, the dampening can be enlarged
+        if (res<res_old) {
+            // update x and F with x_try and F_try
+            N_VScale(1.0, tdata->x, tdata->x_old);
+            N_VScale(1.0, tdata->xdot, tdata->xdot_old);
+            
+            // Check condition
+            if (res < udata->atol) {
+                // Return number of Newton steps needed
+                rdata->newton_numsteps[ntry-1] = inewt + 1;
+                stat1 = 0;
+                break;
+            }
+            
+            if (stat1 != 0)
+            {
+                // increase dampening
+                gamma = fmax(1.0, 2.0*gamma);
+            
+                // solve the new linear system
+                status = getNewtonStep(udata, rdata, tdata, ami_mem, ntry, inewt, delta);
+            }
+        } else {
+            gamma = gamma/4.0;
+        }
+    }
+
+    // Clean up worksapce
+    N_VDestroy_Serial(delta);
+    N_VDestroy_Serial(rel_x);
+    N_VDestroy_Serial(tmp_x);
+    
+    rdata->newton_numsteps[ntry-1] = udata->newton_maxsteps;
+    
+    return(stat1);
+    
+}
+
+/* ------------------------------------------------------------------------------------- */
+/* ------------------------------------------------------------------------------------- */
+/* ------------------------------------------------------------------------------------- */
+
+int getNewtonStep(UserData *udata, ReturnData *rdata, TempData *tdata, void *ami_mem, int ntry, int nnewt, N_Vector ns_delta) {
+    /**
+     * getNewtonStep acomputes the Newton Step by solving a linear system
+     *
+     * @param[in] ami_mem pointer to the solver memory block @type *void
+     * @param[in] udata pointer to the user data struct @type UserData
+     * @param[in] tdata pointer to the temporary data struct @type TempData
+     * @param[out] tdata pointer to the temporary data struct @type TempData
+     * @return void
+     */
+    
+     int il;
+     int stat1 = -1;
+     double rho;
+     double rho1;
+     double alpha;
+     double beta;
+     double omega;
+     double res;
+     double rel_res;
+    
+     N_Vector ns_p = N_VNew_Serial(udata->nx);
+     N_Vector ns_h = N_VNew_Serial(udata->nx);
+     N_Vector ns_t = N_VNew_Serial(udata->nx);
+     N_Vector ns_s = N_VNew_Serial(udata->nx);
+     N_Vector ns_r = N_VNew_Serial(udata->nx);
+     N_Vector ns_rr = N_VNew_Serial(udata->nx);
+     N_Vector ns_rt = N_VNew_Serial(udata->nx);
+     N_Vector ns_v = N_VNew_Serial(udata->nx);
+     N_Vector ns_Jv = N_VNew_Serial(udata->nx);
+     N_Vector ns_tmp = N_VNew_Serial(udata->nx);
+     N_Vector ns_Jdiag = N_VNew_Serial(udata->nx);
+    
+     N_VScale(-1.0, tdata->xdot, tdata->xdot);
+    
+    // Get the diagonal of the Jacobian for preconditioning
+    fJDiag(tdata->t, ns_Jdiag, tdata->x, udata);
+    
+    // Ensure positivity of entries in ns_Jdiag
+    N_VConst(1.0, ns_p);
```

Comment:
I don't understand what happens with `ns_p`. It it initialized with one, used to scale `ns_tmp` by 1.0 and then set to 0. 

Why replace N_VLinearSum by N_VScale and directly set `ns_p` to 0?